src/Spanner/Transaction.php

```diff
@@ -0,0 +1,472 @@
+     * @param array $data The data to insert or update.
+     * @return Transaction The transaction, to enable method chaining.
+     */
+    public function insertOrUpdate($table, array $data)
+    {
+        return $this->insertOrUpdateBatch($table, [$data]);
+    }
+
+    /**
+     * Enqueue one or more insert or update mutations.
+     *
+     * Example:
+     * ```
+     * $transaction->insertOrUpdateBatch('Posts', [
+     *     [
+     *         'ID' => 10,
+     *         'title' => 'My New Post',
+     *         'content' => 'Hello World'
+     *     ]
+     * ]);
+     * ```
+     *
+     * @param string $table The table to insert into or update.
+     * @param array $dataSet The data to insert or update.
+     * @return Transaction The transaction, to enable method chaining.
+     */
+    public function insertOrUpdateBatch($table, array $dataSet)
+    {
+        $this->enqueue(Operation::OP_INSERT_OR_UPDATE, $table, $dataSet);
+
+        return $this;
+    }
+
+    /**
+     * Enqueue an replace mutation.
+     *
+     * Example:
+     * ```
+     * $transaction->replace('Posts', [
+     *     'ID' => 10,
+     *     'title' => 'My New Post [Replaced]',
+     *     'content' => 'Hello Moon'
+     * ]);
+     * ```
+     *
+     * @param string $table The table to replace into.
+     * @param array $data The data to replace.
+     * @return Transaction The transaction, to enable method chaining.
+     */
+    public function replace($table, array $data)
+    {
+        return $this->replaceBatch($table, [$data]);
+    }
+
+    /**
+     * Enqueue one or more replace mutations.
+     *
+     * Example:
+     * ```
+     * $transaction->replaceBatch('Posts', [
+     *     [
+     *         'ID' => 10,
+     *         'title' => 'My New Post [Replaced]',
+     *         'content' => 'Hello Moon'
+     *     ]
+     * ]);
+     * ```
+     *
+     * @param string $table The table to replace into.
+     * @param array $dataSet The data to replace.
+     * @return Transaction The transaction, to enable method chaining.
+     */
+    public function replaceBatch($table, array $dataSet)
+    {
+        $this->enqueue(Operation::OP_REPLACE, $table, $dataSet);
+
+        return $this;
+    }
+
+    /**
+     * Enqueue an delete mutation.
+     *
+     * Example:
+     * ```
+     * $keySet = new KeySet([
+     *     'keys' => [10]
+     * ]);
+     *
+     * $transaction->delete('Posts', $keySet);
+     * ```
+     *
+     * @param string $table The table to mutate.
+     * @param KeySet $keySet The KeySet to identify rows to delete.
+     * @return Transaction The transaction, to enable method chaining.
+     */
+    public function delete($table, KeySet $keySet)
+    {
+        $this->enqueue(Operation::OP_DELETE, $table, [$keySet]);
+
+        return $this;
+    }
+
+    /**
+     * Roll back a transaction.
+     *
+     * Rolls back a transaction, releasing any locks it holds. It is a good idea
```

Comment:
Maybe the wording should be stronger here.` It must be called for any transaction that includes one or more read or query requests and decides not to commit.`